test/manipulation.test.js

```diff
@@ -1044,6 +1044,278 @@ describe('manipulation', function() {
+          { id: 10, name: 'Ritz', city: undefined },
+          function(err, instance) {
+            if (err) return done(err);
+            instance.toObject().should.have.properties({
+              id: 10,
+              name: 'Ritz',
+              city: undefined,
+            });
+
+            Person.upsertWithWhere({ id: 10 },
+                { name: 'updated name' },
+                function(err, updated) {
+                  if (err) return done(err);
+                  var result = updated.toObject();
+                  result.should.have.properties({
+                    id: instance.id,
+                    name: 'updated name',
+                  });
+                  should.equal(result.city, null);
+                  done();
+                });
+          });
+    });
+
+    it('updates specific instances when PK is not an auto-generated id', function(done) {
+      Person.create([
+        { name: 'nameA', city: 'cityA' },
+        { name: 'nameB', city: 'cityB' },
+      ], function(err, instance) {
+        if (err) return done(err);
+
+        Person.upsertWithWhere({ name: 'nameA' },
+            { city: 'newCity' },
+            function(err, instance) {
+              if (err) return done(err);
+
+              var result = instance.toObject();
+              result.should.have.properties({
+                name: 'nameA',
+                city: 'newCity',
+              });
+
+              Person.find(function(err, persons) {
+                if (err) return done(err);
+
+                persons.should.have.length(3);
+                persons[1].name.should.equal('nameA');
+                persons[1].city.should.equal('newCity');
+                persons[2].name.should.equal('nameB');
+                persons[2].city.should.equal('cityB');
+                done();
+              });
+            });
+      });
+    });
+
+    it('should allow save() of the created instance', function(done) {
+      Person.upsertWithWhere({ id: 999 },
+          { name: 'a-name' },
+          function(err, inst) {
+            if (err) return done(err);
+            inst.save(done);
+          });
+    });
+
+    it('works without options on create (promise variant)', function(done) {
+      var person = { id: 123, name: 'a', city: 'city a' };
+      Person.upsertWithWhere({ id: 123 }, person)
+          .then(function(p) {
+            should.exist(p);
+            p.should.be.instanceOf(Person);
+            p.id.should.be.equal(person.id);
+            p.should.not.have.property('_id');
+            p.name.should.equal(person.name);
+            p.city.should.equal(person.city);
+            return Person.findById(p.id)
+                .then(function(p) {
+                  p.id.should.equal(person.id);
+                  p.id.should.not.have.property('_id');
+                  p.name.should.equal(person.name);
+                  p.city.should.equal(person.city);
+                  done();
+                });
+          })
```

Comment:
Nitpick; we , LoopBack team uses 2 indentations, instead of 4 indentations; like the following
```
    .then(function(p) {
      p.id.should.equal(person.id);
      p.id.should.not.have.property('_id');
      p.name.should.equal(person.name);
      p.city.should.equal(person.city);
      done();
    });
```